airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java

```diff
@@ -227,26 +232,229 @@ private Builder<String, Object> generateMetadata(final StandardSync standardSync
+        default -> {
+          // shouldn't be possible to reach this case
+          throw new RuntimeException("Unrecognized scheduleType!");
+        }
+      }
+    }
+  }
+
+  /**
+   * Given an existing AirbyteCatalog and a patch AirbyteCatalog, return a new catalog consisting of
+   * all streams from the patch catalog, plus any remaining streams from the existing catalog.
+   * Preserves the order of any streams that were present in the existing catalog, and adds new
+   * streams from the patch in the order they were provided.
+   */
+  private static AirbyteCatalog toPatchedAirbyteCatalog(final AirbyteCatalog existing, final AirbyteCatalog patch) {
+
+    final Map<StreamDescriptor, AirbyteStreamAndConfiguration> existingDescriptorToStreamAndConfig = getDescriptorToStreamAndConfigMap(existing);
+    final Map<StreamDescriptor, AirbyteStreamAndConfiguration> patchDescriptorToStreamAndConfig = getDescriptorToStreamAndConfigMap(patch);
+
+    final Map<StreamDescriptor, AirbyteStreamAndConfiguration> merged =
+        Stream.of(existingDescriptorToStreamAndConfig, patchDescriptorToStreamAndConfig)
+            .flatMap(map -> map.entrySet().stream())
+            .collect(Collectors.toMap(
+                Map.Entry::getKey,
+                Map.Entry::getValue,
+                // if a stream is present in both the existing map and the patch map, keep the patched value
+                (existingVal, patchedVal) -> patchedVal));
+
+    final List<AirbyteStreamAndConfiguration> sortedMergedStreams = sortStreamsForPatchedCatalog(merged.values().stream().toList(), existing, patch);
+
+    return new AirbyteCatalog().streams(sortedMergedStreams);
+  }
+
+  private static Map<StreamDescriptor, AirbyteStreamAndConfiguration> getDescriptorToStreamAndConfigMap(final AirbyteCatalog catalog) {
+    return catalog.getStreams()
+        .stream()
+        .collect(Collectors.toMap(
+            s -> getStreamDescriptorForStream(s.getStream()),
+            s -> s));
+  }
+
+  private static StreamDescriptor getStreamDescriptorForStream(final AirbyteStream stream) {
+    return new StreamDescriptor().name(stream.getName()).namespace(stream.getNamespace());
+  }
+
+  private static Map<StreamDescriptor, Integer> getStreamPositionMap(final AirbyteCatalog catalog) {
+    final Map<StreamDescriptor, Integer> positions = new HashMap<>();
+    for (int i = 0; i < catalog.getStreams().size(); i++) {
+      positions.put(getStreamDescriptorForStream(catalog.getStreams().get(i).getStream()), i);
+    }
+    return positions;
+  }
+
+  /**
+   * Given an unsorted list of streams, the original catalog, and the patch catalog, returns a sorted
+   * list of streams where existing stream positions are preserved, and new streams from the patch are
+   * added in the order they were listed in the patch.
+   */
+  private static List<AirbyteStreamAndConfiguration> sortStreamsForPatchedCatalog(final List<AirbyteStreamAndConfiguration> mergedStreams,
```

Comment:
@davinchia want to call attention to this logic, which is new since your last review. Basically, now that callers don't need to supply the entire catalog for every request, I want to make sure we preserve the ordering of the streams when patching the catalog. This helper is pretty verbose but I think it gets the job done

Reply:
Thanks!

Reply:
Remind me, why do we care about preserving the order?

Reply:
I don't think it's _strictly_ necessary, but I think it shows some user empathy to preserve the order when making changes (much easier to verify the change as a human looking at an update response, no unexpected order changes in UIs that show the catalog, etc) and it's also easier to write tests since we can use equality matchers on entire `ConnectionRead` objects, rather than needing to implement a custom matcher that allows the catalog to be unsorted, but wants equality for every other property.

Reply:
Makes sense. Can we leave a comment here explaining that? Want to make it clear this is a UX decision and not an application constraint.